File: packages/crypto/src/identity/keys.ts
import { generateEd25519KeyPair, generateX25519KeyPair, sign, type KeyPair } from '../primitives';

/**
 * The identity key is a **pair of key pairs** — one Ed25519 (signing) and one X25519
 * (Diffie-Hellman) — rather than a single Curve25519 key reused for both via XEdDSA
 * conversion. The X3DH spec explicitly allows either approach (§2.1: "if it's not
 * important to save the extra 32 bytes... using distinct signing and DH keys is
 * simpler"); distinct keys avoid needing an Edwards↔Montgomery conversion step in
 * this codebase, at the cost of an extra 32 public + 32 private bytes per identity —
 * a rounding error against everything else this system already stores per device.
 */
export interface IdentityKeyPair {
  signing: KeyPair; // Ed25519 — signs signed-prekeys, verified by peers before trusting them
  agreement: KeyPair; // X25519 — used directly in X3DH's DH1 step
}

export function generateIdentityKeyPair(): IdentityKeyPair {
  return {
    signing: generateEd25519KeyPair(),
    agreement: generateX25519KeyPair(),
  };
}

export interface SignedPreKey {
  keyId: number;
  keyPair: KeyPair; // X25519
  signature: Uint8Array; // Ed25519 signature over keyPair.publicKey, by the identity signing key
}

/**
 * Rotated periodically by the client (docs/05-crypto-architecture.md's key
 * hierarchy) — a fresh one replaces the old, the old is kept server-side briefly for
 * sessions already in flight, then garbage-collected (server-side concern, not this
 * package's).
 */
export function generateSignedPreKey(identitySigningPrivateKey: Uint8Array, keyId: number): SignedPreKey {
  const keyPair = generateX25519KeyPair();
  const signature = sign(identitySigningPrivateKey, keyPair.publicKey);
  return { keyId, keyPair, signature };
}

export interface OneTimePreKey {
  keyId: number;
  keyPair: KeyPair; // X25519
}

/**
 * Each one-time pre-key is consumed exactly once by whoever claims it first (server
 * enforces this via `one_time_pre_keys.claimed_at`, docs/02-database-schema.md) —
 * this function just generates a batch; the caller is responsible for keeping the
 * private halves until used and then deleting them (see session/session.ts's
 * inbound-session handling, which deletes the matching private key immediately
 * after a successful X3DH derivation).
 */
export function generateOneTimePreKeys(count: number, startKeyId: number): OneTimePreKey[] {
  return Array.from({ length: count }, (_, i) => ({
    keyId: startKeyId + i,
    keyPair: generateX25519KeyPair(),
  }));
}

File: apps/mobile/lib/crypto/identity/keys.dart
/// Direct port of `packages/crypto/src/identity/keys.ts` — see that file's docstring
/// for why identity is a pair of key pairs (Ed25519 signing + X25519 agreement)
/// rather than one XEdDSA-converted key.
library;

import 'dart:typed_data';
import '../primitives.dart';

class IdentityKeyPair {
  final KeyPairBytes signing; // Ed25519
  final KeyPairBytes agreement; // X25519
  const IdentityKeyPair({required this.signing, required this.agreement});
}

Future<IdentityKeyPair> generateIdentityKeyPair() async {
  final signing = await generateEd25519KeyPair();
  final agreement = await generateX25519KeyPair();
  return IdentityKeyPair(signing: signing, agreement: agreement);
}

class SignedPreKey {
  final int keyId;
  final KeyPairBytes keyPair; // X25519
  final Uint8List signature; // Ed25519 signature over keyPair.publicKey
  const SignedPreKey({required this.keyId, required this.keyPair, required this.signature});
}

/// Rotated periodically by the client — a fresh one replaces the old (server-side
/// concern, not this file's).
Future<SignedPreKey> generateSignedPreKey(Uint8List identitySigningPrivateKey, int keyId) async {
  final keyPair = await generateX25519KeyPair();
  final signature = await sign(identitySigningPrivateKey, keyPair.publicKey);
  return SignedPreKey(keyId: keyId, keyPair: keyPair, signature: signature);
}

class OneTimePreKey {
  final int keyId;
  final KeyPairBytes keyPair; // X25519
  const OneTimePreKey({required this.keyId, required this.keyPair});
}

/// Each one-time pre-key is consumed exactly once (server enforces via
/// `one_time_pre_keys.claimed_at`) — this just generates a batch; the caller keeps
/// the private halves until used, then deletes them (see crypto/sessions.dart's
/// inbound-session handling).
Future<List<OneTimePreKey>> generateOneTimePreKeys(int count, int startKeyId) async {
  final out = <OneTimePreKey>[];
  for (var i = 0; i < count; i++) {
    out.add(OneTimePreKey(keyId: startKeyId + i, keyPair: await generateX25519KeyPair()));
  }
  return out;
}
